### PR TITLE
[CALCITE-3668] VolcanoPlanner does not match all the RelSubSet in matchRecursive

### PR DESCRIPTION
**Pull request by hsyuan**
JIRA: https://issues.apache.org/jira/browse/CALCITE-3668

The change was introduced by https://github.com/apache/calcite/pull/784, which I think is wrong.

### CODE REVIEW

**Comment by vlsi** [comment]
@hsyuan , have you seen this discussion?
https://github.com/apache/calcite/pull/784#discussion_r245797826

Can you provide an explanation and/or a test case to elaborate on why the current implementation is not right?

**Comment by hsyuan** [comment]
@vlsi I have updated with a test case, please take a look.

File: core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleCall.java [hunk 1]
```diff
@@ -317,8 +317,7 @@ private void matchRecurse(int solve) {
           final RelSubset subset =
               (RelSubset) inputs.get(operand.ordinalInParent);
           if (operand.getMatchedClass() == RelSubset.class) {
-            // If the rule wants the whole subset, we just provide it
-            successors = ImmutableList.of(subset);
+            successors = subset.getSubsets();
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleCall.java]: The naming is hard.
`subset.getSubsets()` looks confusing :(

**Comment by vlsi** [comment]
@hsyuan , this looks good to me modulo `subset.getSubsets()`.

I think it would be extremely confusing, especially for those who don't know how `RelSet` is different from `RelSubset`.

Even though the method is not in public API, I would suggest picking a name that would convey the meaning.
I don't know. `subset.getSiblingSubsetsWithMatchingTraits()`?

**Comment by hsyuan** [comment]
Agree, the name is confusing. I don't think it will be used by any other places. How about we just drop the method, use this directly:
```
successors = subset.set.subsets.stream()
        .filter(s -> s.getTraitSet().satisfies(subset.getTraitSet()))
        .collect(Collectors.toList());
```

**Comment by vlsi** [comment]
LGTM, thanks

**Pull request merged**